Cover block: Allow setting the height from the placeholder state.

ramonjd opened the pull request:
## Description

This PR allows us to set the min-height value for the Cover Block by resizing the placeholder. 🪄 

It's part of a [wider issue to improve the Cover Block UX](https://github.com/WordPress/gutenberg/issues/28385).

This PR also updates the copy in the placeholder to describe this functionality. Note, there's a [PR](https://github.com/WordPress/gutenberg/pull/34970) to update the Cover Block placeholder copy, so we don't have to do it here.

https://user-images.githubusercontent.com/6458278/135028287-6131ce47-c74d-42b9-8821-5b3793c0bc0b.mp4

Kudos to @jasmussen for the original [UX idea](https://github.com/WordPress/gutenberg/issues/34706#issuecomment-921534880). 🎉 

Resolves https://github.com/WordPress/gutenberg/issues/34706

## Testing

1. Add a Cover Block 
2. Adjust placeholder's height using the resize handle
3. The min-height value in the block inspector controls (in the Dimensions panel) should reflect the resized placeholder's min-height
4. Ensure that the placeholder's background color and border is otherwise unaffected by window resizing.
5. Check that you can interact with all controls and text inside the placeholder, and also drag an image on top of it.
5. Also make sure that the resizer itself only appears for large placeholders. You can test this by resizing the window or inserting a cover block into a narrow column. You shouldn't be able to resize placeholders with a width of `< 480px`
5. Select a color or a media file
6. Ensure that the min-height value is retained and displays correctly in the editor and the frontend.
7. Check that the resizer works as expected, with a min-height of 50px after we've inserted the Cover Block.
8. Things should work as above on a mobile device as well.


## Types of changes
UX enhancement on a block placeholder.

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

ramonjd committed:
initial commit

ramonjd committed:
Added some copy.

ramonjd committed:
Ensure that the resizeable placeholder has a consistent background color (white) and border.

andrewserong commented on packages/block-library/src/cover/edit.js:
```diff
@@ -629,10 +635,25 @@ function CoverEdit( {
+							setAttributes( { minHeight: newMinHeight } );
+							setTemporaryMinHeight( null );
+						} }
+						showHandle={ isSelected }
```
I was wondering if it'd be worth adding the `minHeight={ 250 }` prop to this `ResizableCover` so that it gets passed down to ResizableBox? While values less than 250 are valid for the cover block, it looked a little odd being able to drag the control over the placeholder area. This shouldn't affect the ability to have a smaller height once a background colour is chosen 🤞

| Before | After | 
| --- | --- |
| ![Kapture 2021-09-28 at 16 50 44](https://user-images.githubusercontent.com/14988353/135037436-19ad83ca-66f3-47f5-8ba5-c044e0a9728f.gif) | ![Kapture 2021-09-28 at 16 49 02](https://user-images.githubusercontent.com/14988353/135037207-8cd998df-7d5e-4064-b2cd-fa146a6a57e4.gif) |

ramonjd replied:
Good idea. I'll have a play around with your suggestion.

andrewserong commented on packages/block-library/src/cover/editor.scss:
```diff
@@ -116,3 +117,12 @@
+.wp-block-cover.is-placeholder .cover-block__cover-placeholder-container {
+	background: $white;
+	width: 100%;
+	.block-editor-media-placeholder {
```
Tiniest of nits: it looks like the `box-shadow` comes from the `.components-placeholder` class, but the element here also happens to have the `.block-editor-media-placeholder` class name. Would it be worth switching to the other class name for consistency, or is it better to keep it as-is for specificity?

jasmussen replied:
This will need massaging as well, as we need the resting and focus states to be the same as other placeholders, which very probably requires we use a box shadow after all. Happy to help if this turns gnarly.

ramonjd replied:
Thanks for noticing this! I'll work on it and get back to you if the gnarliness defeats me! :)

andrewserong reviewed (commented):
Nice work @ramonjd! This is testing well for me, and it's great being able to interact with the minimum height while in the placeholder state. I've left a comment about whether it'd be worth constraining the minimum height in the placeholder state to leave clearance for the placeholder, and a tiny comment about the classnames (but feel free to ignore if it isn't relevant).

The wording looks good to me, but CC: @apeatling since he's been working on the placeholder wording, too.

jasmussen commented:
Nice, this is working better than I had expected. I was happy to see the min height also being present in the inspector. Here's a GIF showing the cover resize:
![cover resize](https://user-images.githubusercontent.com/1204802/135040409-3fe09b43-53a4-4ac4-ab7e-9cf3db10ccda.gif)

A few small things:

- The focus style is now off, so you see both the black border and the blue outline. This is likely due to the addition of the separate black border. I can help you figure this out if need be. You can compare with the Image placeholder and the style of border should be the same.
- The 2px border radius on the placeholder is missing.

Some of the larger stuff, Andrew suggests adding a min-height while it's still in a placeholder state. That _might_ be a good temporary solution, but it doesn't feel like a longer term solution to have separate min-heights depending on state. One thing that might or might not work would be to enhance the resizeobserver system that's built into the `Placeholder` component. Right now it applies `is-small`, `is-medium` and `is-large` classes to the placeholder so that when they are inserted in narrow contexts (such as a column) we can hide descriptions or otherwise minimize the content of the placeholders. If this were built right, we could potentially minimize the placeholder both vertically and horizontally to the point that the entire thing just becomes an upload button, a la the Site Logo placeholder (#35096). It might simply be enough to apply the`is-small` CSS class not just based on the width, but based on the height as well. Can you take a quick look and see if enhancing this with height awareness would be easy or complex?

apeatling reviewed (commented):
This works well for me. I do agree we need some way to stop the resizable placeholder from sizing smaller than the contents of the placeholder.

![2021-09-28 10 03 13](https://user-images.githubusercontent.com/1464705/135132733-82dbf0cd-a553-444a-a09e-0a953e7e1ebe.gif)

ramonjd commented:
Thanks @andrewserong @apeatling and @jasmussen for the feedback. I didn't catch the min-height of the container and the focus states, so it's really helpful. I'll work on your suggestions and aim to get a revision up soon.

ramonjd committed:
Experiment to add vertical classes to the placeholder and use it to determine the min-height of the resizeable placeholder.

ramonjd committed:
limit the resize behaviour to `is-large` placeholders

Reverting changes to the Placeholder `useResizeObserver` since it was a test.
limit the resize behaviour to `is-large` placeholders

ramonjd commented:
I've converted this PR back to draft in order to experiment.

Firstly, sorry for the wall of text and images. I thought it best to document my progress and gather feedback here.

Secondly, it turns out most of the style/outline/focus issues can be resolved by moving the ResizableCover component inside the CoverPlaceholder component. 🎉 

> Some of the larger stuff, Andrew suggests adding a min-height while it’s still in a placeholder state. That might be a good temporary solution, but it doesn’t feel like a longer term solution to have separate min-heights depending on state.

Actually, having played around with things I agree. For example, giving the resizer a static min-height of `250` requires us also to set the `.components-placeholder` min-height to `250`. 

This is so the initial state of both matches up and the resize handler knows where to position itself, and there's no overlap. Here's what happens when we don't do this and insert a Cover Block:

<img width="683" alt="Screen Shot 2021-09-30 at 11 46 18 am" src="https://user-images.githubusercontent.com/6458278/135380653-109b4ecf-7fca-4914-b70e-1315340bbd02.png">

Furthermore, a value of `250` (or whatever it is) is fairly arbitrary and doesn’t take into account the contents of the placeholder.  

So for example, we might set `250`, which works for now in for the large placeholder, but later content might be added which will bring the height of the placeholder above 250.

To demonstrate what I'm talking about, here's what happens when we set the min-height to `200`:

![min-height-cover-200](https://user-images.githubusercontent.com/6458278/135380788-63f3c762-a91e-4345-a79d-86417279207a.gif)

> Can you take a quick look and see if enhancing this with height awareness would be easy or complex?

Thanks for suggesting this. 🙇 

So observing the height as well as the width in the [Placeholder component](https://github.com/WordPress/gutenberg/blob/a7e2895829c16ecd77a5ba22d84f1dee1cfb0977/packages/components/src/placeholder/index.js#L42) is trivial using the `react-resize-aware` library – 

```js
const [ resizeListener, { width, height } ] = useResizeObserver();

let modifierClassNames;
if ( typeof width === 'number' ) {
	modifierClassNames = {
                 // TODO rename these classes, e.g., `is-width-large`, `is-height-large` etc
		'is-large': width >= 480, 
		'is-medium': width >= 160 && width < 480,
		'is-small': width < 160,
                 // TODO determine the most appropriate breakpoint values for height
		'is-tall': height >= 250,
		'is-middle': height >= 100 && height < 250,
		'is-short': height < 100,
	};
}
```

– it's how we apply it to this specific use case.

Using a combination of the vertical and horizontal classes we might pull off a consistent experience of the following:

![reveal-cover](https://user-images.githubusercontent.com/6458278/135389407-55fa98a0-e076-49eb-ad40-16c49ec208d9.gif)

It's a promising direction. 

The placeholders might be taller or shorter depending on the content however, so it's been challenging to reconcile where and when we hide and show elements. 

Before I dive down that rabbit hole, it would be great to know if I've understood your comments @jasmussen 😆 

Another option (the one I've just committed), is one that allows us to test the behaviour with a smaller footprint, would be to limit the resize behaviour to `is-large` placeholders for now, e.g., 

```css
	.components-resizable-box__container {
		display: none;
	}
	.components-placeholder {
		&.is-large {
			min-height: 250px;
			.components-resizable-box__container {
				min-height: 250px;
				display: block;
			}
		}
```

Here's what that would look like:

![only-large-cover-block](https://user-images.githubusercontent.com/6458278/135383341-242a193f-f3c8-46b0-8ead-66a96f11bd77.gif)

The benefit would be that we expose the functionality, and then work in parallel on exposing vertical Placeholder classes.

jasmussen commented:
> So observing the height as well as the width in the Placeholder component is trivial using the react-resize-aware library –

I like the sound of that.

Instead of creating separate cases for width and height, though, I think it might be worth it to just lump them together as one, to keep things simple. So instead of:

```
if ( typeof width === 'number' ) {
	modifierClassNames = {
                 // TODO rename these classes, e.g., `is-width-large`, `is-height-large` etc
		'is-large': width >= 480, 
		'is-medium': width >= 160 && width < 480,
		'is-small': width < 160,
                 // TODO determine the most appropriate breakpoint values for height
		'is-tall': height >= 250,
		'is-middle': height >= 100 && height < 250,
		'is-short': height < 100,
	};
}
```

maybe something like:

```
if ( typeof width === 'number' ) {
	modifierClassNames = {
		'is-large': width >= 480 || height >=250, 
		'is-medium': width >= 160 && width < 480 || height >= 100 && height < 250,
		'is-small': width < 160 || height < 100,
	};
}
```

On the one hand, many third party blocks might be leveraging the existing classes, so it's risky to rename them. Additionally, by simply having at most three sizes to handle in the placeholder, it makes it easier to handle the responsive behavior across all placeholders, which is important if we are to keep using it.

I also think, separate from this PR, that we can enhance the design of these placeholder components across the board, and in doing so find a better design for the minimal version of placeholders. For example, [here's what a minimal site logo placeholder could look like](https://github.com/WordPress/gutenberg/issues/35096#issuecomment-926631753), and [here's a work in progress exploration on adding the color control to the Cover block toolbar](https://github.com/WordPress/gutenberg/issues/35093#issuecomment-928933121), which would at least make it available in the smaller breakpoints.

ramonjd commented:
> On the one hand, many third party blocks might be leveraging the existing classes, so it's risky to rename them.

Definitely. What's more, I noticed that Cover Blocks within narrow columns for example, can be both very tall but narrow, so it'd be a difficult one to categorize.

It might be worth an experimental PR just to see what's possible with various placeholder dimensions in light of the work you've linked to (thank you for that too!).

jasmussen commented:
> It might be worth an experimental PR just to see what's possible with various placeholder dimensions in light of the work you've linked to (thank you for that too!).

There's definitely work to do, including designwise. But in light of that, it's also good to keep things as simple as possible, as simplicity is definitely what we'll seek more of for any iterations!

Nice work.

apeatling reviewed (approved):
Working well, LGTM :+1:

ramonjd committed:
Ensure the placeholder takes up the maximium available width set by the containing block

ramonjd commented:
More testing reveals that the resizer is sitting on top of the controls, meaning that we can't interact with the upload/media library buttons.

<img width="661" alt="Screen Shot 2021-10-08 at 1 15 07 pm" src="https://user-images.githubusercontent.com/6458278/136487666-de073c72-5845-48bf-a7f3-4206a0874669.png">

I suspect the reordering of the resizeable placeholder component in e97b6937536b4087248db7717f73c1b2e0edd987

Investigating...

ramonjd committed:
We need to reshuffle the order of the elements to ensure that:

- a user can interact with the placeholder controls
- we can target the resizer as a sibling of `.components-placeholder.is-large`

Also cleaning up the CSS and removing unused styles.

ramonjd commented:
@apeatling Thanks for the review! 

I've reordered the markup and added a z-index to the placeholder so that we can interact with the controls this time 😄 

Everything should work as before if you get time to do another round of reviewing. 🙇

jasmussen commented:
Visually this is looking good, and it's a nice feature. However I'm seeing a 250px min-height on the placeholder:
<img width="767" alt="Screenshot 2021-10-08 at 09 25 42" src="https://user-images.githubusercontent.com/1204802/136515977-a36a3272-e739-471d-aeff-f8984f737766.png">

One problem that exists with enforcing a min-height just to accommodate the big placeholder, is that it doesn't account for the `is-medium` and `is-small` breakpoints. In those cases even now (try putting a cover in column 1 of a 3 column setup) font sizes are reduced, the description is hidden, and things get smaller. The 250px min-height is invisible there because the color swatches are still shown. But that's arguably a bug, and the color swatches should be hidden at states less than is-large, at which point the min-height becomes problematic.

250px is also taller than the 220 ish px that seem needed in my testing. It seems like we could possibly land this by reducing that min-height.

However I think it's worth doing or ticketing a followup to revisit the cover placeholder and enhance its narrow-context behavior. It seems like with nice flex rules, we should be able to reduce the required min-height all the way down to about 72px: 

<img width="742" alt="Screenshot 2021-10-08 at 09 29 11" src="https://user-images.githubusercontent.com/1204802/136516180-29643a31-3b36-446e-b1d5-5506d8d3cd8f.png">

_Bad inspector mockup, we'd of course want things to align._

ramonjd commented:
> One problem that exists with enforcing a min-height just to accommodate the big placeholder, is that it doesn't account for the is-medium and is-small breakpoints. 

True. That's why we restricted the resizer to show only for `is-large` placeholders. Not an ideal long term solution, I know, but it's the only way I could think of getting around the constraints you mentioned.

> 250px is also taller than the 220 ish px that seem needed in my testing. It seems like we could possibly land this by reducing that min-height.

WIll do! Thanks! 🙇 

> However I think it's worth doing or ticketing a followup to revisit the cover placeholder and enhance its narrow-context behavior. It seems like with nice flex rules, we should be able to reduce the required min-height all the way down to about 72px:

For sure. I think we should be able to find a way to have the placeholder content respond to the container's dimensions. I really only spent an hour or two hacking at it. How that plays with the resizer, assuming we unleash it across medium and small dimensions is well is to be seen 😄

ramonjd commented:
> 250px is also taller than the 220 ish px that seem needed in my testing.

The copy is new so it's still untranslated, but I think the 250px was originally set to accommodate two lines of text, which might be required for German or Russian for example. 

At 220px we get a bit of overlap:


![Oct-08-2021 19-26-39](https://user-images.githubusercontent.com/6458278/136524049-063ba946-0d46-4b25-af35-a15311e2b78d.gif)

It's not a huge eyesore I think. 240px seems to be the spot where we don't see this effect if we want to avoid it.

It does expose fragility in that we're relying on unknown content height. We should try to overcome this limitation in future explorations. 👍 

Cheers!

jasmussen commented:
> It does expose fragility in that we're relying on unknown content height. We should try to overcome this limitation in future explorations. 👍

It seems fine to hide elements, including the description and color swatches, as the placeholder gets resized, same as we already do for horizontal narrowing. So long as we come back to this, probably just pick a value and go!

ramonjd committed:
Bumping min-height of is-large placeholder resizer down to 240px

ramonjd committed:
Because we removed `minHeight={ COVER_MIN_HEIGHT }` on the resizeable container so that our changes in editor.scss take precedence, we have to reinstate the COVER_MIN_HEIGHT of 50px now via css

andrewserong reviewed (approved):
Nice work @ramonjd! This is testing nicely for me, and aside from the additional control while selected, the placeholder state looks consistent with other placeholders like the Image block. Like Joen, I noticed that the min height of the placeholder is now a little taller, however that also seems to help things in smaller viewport widths where the descriptive text wraps on a second line, so I think the extra breathing room there seems worth it (as you mentioned with factoring in the potential length of translations, too).

✅ Min height control in placeholder state updates values correctly
✅ Control is not shown on smaller Cover placeholder state when inserted within a Column block
✅ Placeholder's background and border colors are as they were before
✅ Interacting within the Cover block is working as before (selecting color, image, dragging an image from the desktop onto the placeholder, etc)
✅ Tested on Safari, FF, Chrome, and Edge on Mac and Safari and Chrome on iOS (iPhone 8)

![Kapture 2021-10-12 at 14 34 38](https://user-images.githubusercontent.com/14988353/136887303-bb03c7a4-65b7-454c-8c9a-1462b133199c.gif)

LGTM! 🎉

ramonjd commented:
Thanks again the thorough review @andrewserong 🙇
